src/Npgsql/ConnectorPool.cs

```diff
@@ -379,15 +379,15 @@ internal void Clear()
             var count = _idleCount;
             while (count > 0 && _idleConnectorReader.TryRead(out var connector))
             {
-                if (CheckIdleConnector(connector))
+                if (CheckIdleConnector(connector, canDelete: false))
                 {
-                    CloseConnector(connector);
+                    CloseConnector(connector, canDelete: false);
                     count--;
                 }
             }
         }
 
-        void CloseConnector(NpgsqlConnector connector)
+        void CloseConnector(NpgsqlConnector connector, bool canDelete = true)
```

Comment:
nit: since canDelete is only true when calling this from pruning, I'd inverse the flag (default to false).